File: VP-ObjC/CodexUsageReader.m
/* audience: machine */
#import "CodexUsageReader.h"

@implementation VPCodexUsageReader

+ (NSString *)codexDir {
    return [NSHomeDirectory() stringByAppendingString:@"/.codex"];
}

+ (NSString *)authPath {
    return [[self codexDir] stringByAppendingString:@"/auth.json"];
}

+ (BOOL)isInstalled {
    BOOL isDir = NO;
    BOOL exists = [[NSFileManager defaultManager] fileExistsAtPath:[self codexDir] isDirectory:&isDir];
    return exists && isDir;
}

+ (nullable NSDictionary *)loadCredential {
    NSData *data = [NSData dataWithContentsOfFile:[self authPath]];
    if (!data) return nil;
    id obj = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
    if (![obj isKindOfClass:[NSDictionary class]]) return nil;
    NSDictionary *tokens = ((NSDictionary *)obj)[@"tokens"];
    if (![tokens isKindOfClass:[NSDictionary class]]) return nil;
    NSString *accessToken = tokens[@"access_token"];
    if (![accessToken isKindOfClass:[NSString class]] || accessToken.length == 0) return nil;
    return tokens;
}

+ (void)pollWithCompletion:(void (^)(VPLimitRow *_Nullable))completion {
    if (![self isInstalled]) { completion(nil); return; }
    NSDictionary *credential = [self loadCredential];
    if (!credential) { completion(nil); return; }

    NSString *accessToken = credential[@"access_token"];
    NSString *accountID = credential[@"account_id"] ?: @"";

    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"https://chatgpt.com/backend-api/wham/usage"]
                                                             cachePolicy:NSURLRequestReloadIgnoringLocalCacheData
                                                         timeoutInterval:15];
    [request setValue:[NSString stringWithFormat:@"Bearer %@", accessToken] forHTTPHeaderField:@"Authorization"];
    [request setValue:accountID forHTTPHeaderField:@"ChatGPT-Account-Id"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];

    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request
        completionHandler:^(NSData *_Nullable data, NSURLResponse *_Nullable response, NSError *_Nullable error) {
            NSInteger status = [(NSHTTPURLResponse *)response statusCode];
            if (status < 200 || status >= 300 || data == nil) {
                dispatch_async(dispatch_get_main_queue(), ^{ completion(nil); });
                return;
            }
            id payload = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
            if (![payload isKindOfClass:[NSDictionary class]]) {
                dispatch_async(dispatch_get_main_queue(), ^{ completion(nil); });
                return;
            }
            id rateLimit = ((NSDictionary *)payload)[@"rate_limit"];
            id window = [rateLimit isKindOfClass:[NSDictionary class]] ? ((NSDictionary *)rateLimit)[@"primary_window"] : nil;
            if (![window isKindOfClass:[NSDictionary class]]) {
                dispatch_async(dispatch_get_main_queue(), ^{ completion(nil); });
                return;
            }
            NSDictionary *windowDict = (NSDictionary *)window;
            NSNumber *pctNum = windowDict[@"used_percent"];
            if (![pctNum isKindOfClass:[NSNumber class]]) {
                dispatch_async(dispatch_get_main_queue(), ^{ completion(nil); });
                return;
            }
            NSDate *resetsAt = nil;
            NSNumber *resetAt = windowDict[@"reset_at"];
            NSNumber *resetAfterSeconds = windowDict[@"reset_after_seconds"];
            if ([resetAt isKindOfClass:[NSNumber class]]) {
                resetsAt = [NSDate dateWithTimeIntervalSince1970:resetAt.doubleValue];
            } else if ([resetAfterSeconds isKindOfClass:[NSNumber class]]) {
                resetsAt = [NSDate dateWithTimeIntervalSinceNow:resetAfterSeconds.doubleValue];
            }
            VPLimitRow *row = [[VPLimitRow alloc] initWithLabel:@"Sesión actual" usedPercent:pctNum.doubleValue resetsAt:resetsAt];
            dispatch_async(dispatch_get_main_queue(), ^{ completion(row); });
        }];
    [task resume];
}

@end

File: VP/CodexUsageReader.swift
// audience: machine
// Codex is not installed on this Mac, so this reads the same on-disk
// convention Sources/Providers/CodexProfile.swift and CodexCredentials.swift
// use (~/.codex/auth.json, tokens.{access_token,account_id}) and, only if
// that folder exists, calls the same endpoint CodexLocalProvider.swift does
// (chatgpt.com/backend-api/wham/usage, rate_limit.primary_window). If the
// folder is missing this never touches the network and reports
// `.notInstalled` so the ring can grey out honestly instead of inventing data.
import Foundation

enum CodexUsageReader {
    private static var home: String { NSHomeDirectory() }
    private static var codexDir: String { home + "/.codex" }
    private static var authPath: String { codexDir + "/auth.json" }

    static var isInstalled: Bool {
        var isDir: ObjCBool = false
        return FileManager.default.fileExists(atPath: codexDir, isDirectory: &isDir) && isDir.boolValue
    }

    private struct Credential {
        let accessToken: String
        let accountID: String
    }

    private static func loadCredential() -> Credential? {
        struct Auth: Decodable {
            struct Tokens: Decodable {
                let access_token: String
                let account_id: String
            }
            let tokens: Tokens
        }
        guard let data = FileManager.default.contents(atPath: authPath),
              let auth = try? JSONDecoder().decode(Auth.self, from: data),
              !auth.tokens.access_token.isEmpty else { return nil }
        return Credential(accessToken: auth.tokens.access_token, accountID: auth.tokens.account_id)
    }

    private struct Window: Decodable {
        let used_percent: Double?
        let reset_at: Double?
        let reset_after_seconds: Double?
    }
    private struct RateLimit: Decodable {
        let primary_window: Window?
    }
    private struct Response: Decodable {
        let rate_limit: RateLimit?
    }

    /// Only ever called when `isInstalled` is true. Completes with nil on any
    /// failure (no credential, network error, bad response) so the caller can
    /// show "installed but unavailable" rather than crash.
    static func poll(completion: @escaping (LimitRow?) -> Void) {
        guard isInstalled, let credential = loadCredential() else { completion(nil); return }

        var request = URLRequest(
            url: URL(string: "https://chatgpt.com/backend-api/wham/usage")!,
            cachePolicy: .reloadIgnoringLocalCacheData,
            timeoutInterval: 15
        )
        request.setValue("Bearer \(credential.accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue(credential.accountID, forHTTPHeaderField: "ChatGPT-Account-Id")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        URLSession.shared.dataTask(with: request) { data, response, _ in
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard (200..<300).contains(status), let data,
                  let payload = try? JSONDecoder().decode(Response.self, from: data),
                  let window = payload.rate_limit?.primary_window,
                  let pct = window.used_percent else {
                DispatchQueue.main.async { completion(nil) }
                return
            }
            let resetsAt = window.reset_at.map { Date(timeIntervalSince1970: $0) }
                ?? window.reset_after_seconds.map { Date().addingTimeInterval($0) }
            let row = LimitRow(label: "Sesión actual", usedPercent: pct, resetsAt: resetsAt)
            DispatchQueue.main.async { completion(row) }
        }.resume()
    }
}
